File: firebase-firestore/src/test/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProviderTest.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.firestore.auth;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApiNotAvailableException;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.internal.IdTokenListener;
import com.google.firebase.auth.internal.InternalAuthProvider;
import com.google.firebase.firestore.testutil.DelayedDeferred;
import com.google.firebase.firestore.testutil.ImmediateDeferred;
import com.google.firebase.firestore.testutil.UnavailableDeferred;
import com.google.firebase.firestore.util.Listener;
import com.google.firebase.internal.InternalTokenResult;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;
import org.mockito.quality.Strictness;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class FirebaseAuthCredentialsProviderTest {

  @Rule public final MockitoRule mocks = MockitoJUnit.rule().strictness(Strictness.STRICT_STUBS);

  @Mock InternalAuthProvider mockInternalAuthProvider;
  @Mock Listener<User> mockUserListener;
  @Mock InternalTokenResult mockInternalTokenResult;
  @Mock GetTokenResult mockGetTokenResult;
  @Mock GetTokenResult mockGetTokenResult2;
  @Captor ArgumentCaptor<IdTokenListener> idTokenListenerCaptor;

  @Test
  public void setChangeListenerShouldBeCalledWithUnauthenticatedIfProviderIsNotAvailable() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new UnavailableDeferred<>());

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);

    verify(mockUserListener).onValue(User.UNAUTHENTICATED);
  }

  @Test
  public void setChangeListenerShouldBeCalledWithUnauthenticatedIfUidIsNull() {
    when(mockInternalAuthProvider.getUid()).thenReturn(null);
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);

    verify(mockUserListener).onValue(User.UNAUTHENTICATED);
  }

  @Test
  public void setChangeListenerShouldBeCalledWithAuthenticatedUserIfUidIsNotNull() {
    when(mockInternalAuthProvider.getUid()).thenReturn("TestUID");
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);

    verify(mockUserListener).onValue(new User("TestUID"));
  }

  @Test
  public void
      setChangeListenerShouldBeCalledWithUnauthenticatedUserWhenProviderWithNullUidBecomesAvailable() {
    DelayedDeferred<InternalAuthProvider> delayedDeferredInternalAuthProvider =
        new DelayedDeferred<>();
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider);

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);
    verify(mockUserListener).onValue(User.UNAUTHENTICATED);
    when(mockInternalAuthProvider.getUid()).thenReturn(null);
    delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider);

    verify(mockUserListener, times(2)).onValue(User.UNAUTHENTICATED);
  }

  @Test
  public void
      setChangeListenerShouldBeCalledWithAuthenticatedUserWhenProviderWithAuthenticatedUserBecomesAvailable() {
    DelayedDeferred<InternalAuthProvider> delayedDeferredInternalAuthProvider =
        new DelayedDeferred<>();
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider);

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);
    verify(mockUserListener).onValue(User.UNAUTHENTICATED);
    when(mockInternalAuthProvider.getUid()).thenReturn("TestUID");
    delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider);

    verify(mockUserListener).onValue(new User("TestUID"));
  }

  @Test
  public void setChangeListenerShouldBeCalledWhenIdTokenChanges() {
    when(mockInternalAuthProvider.getUid()).thenReturn("TestUID1");
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    verify(mockInternalAuthProvider).addIdTokenListener(idTokenListenerCaptor.capture());

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);
    verify(mockUserListener).onValue(new User("TestUID1"));
    when(mockInternalAuthProvider.getUid()).thenReturn("TestUID2");
    idTokenListenerCaptor.getValue().onIdTokenChanged(mockInternalTokenResult);

    verify(mockUserListener).onValue(new User("TestUID2"));
  }

  @Test
  public void removeChangeListenerShouldStopNotifyingTheListener() {
    DelayedDeferred<InternalAuthProvider> delayedDeferredInternalAuthProvider =
        new DelayedDeferred<>();
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider);

    firebaseAuthCredentialsProvider.setChangeListener(mockUserListener);
    verify(mockUserListener).onValue(User.UNAUTHENTICATED);
    firebaseAuthCredentialsProvider.removeChangeListener();
    delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider);

    verifyNoMoreInteractions(mockUserListener);
  }

  @Test
  public void removeChangeListenerShouldNotThrowIfProviderIsNotAvailable() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new UnavailableDeferred<>());

    firebaseAuthCredentialsProvider.removeChangeListener();
  }

  @Test
  public void removeChangeListenerShouldUnregisterTheIdTokenListener() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    verify(mockInternalAuthProvider).addIdTokenListener(idTokenListenerCaptor.capture());

    firebaseAuthCredentialsProvider.removeChangeListener();

    verify(mockInternalAuthProvider).removeIdTokenListener(idTokenListenerCaptor.getValue());
  }

  @Test
  public void getTokenShouldReturnAFailedTaskIfProviderIsNotAvailable() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new UnavailableDeferred<>());

    Task<String> task = firebaseAuthCredentialsProvider.getToken();

    assertThat(task.isComplete()).isTrue();
    assertThat(task.isSuccessful()).isFalse();
    assertThat(task.getException()).isInstanceOf(FirebaseApiNotAvailableException.class);
  }

  @Test
  public void getTokenShouldReturnAFailedTaskIfGetAccessTokenFails() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    Exception getAccessTokenException = new Exception();
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forException(getAccessTokenException));

    Task<String> task = firebaseAuthCredentialsProvider.getToken();

    assertThat(task.isComplete()).isTrue();
    assertThat(task.isSuccessful()).isFalse();
    assertThat(task.getException()).isSameInstanceAs(getAccessTokenException);
  }

  @Test
  public void getTokenShouldReturnASuccessfulTaskIfGetAccessTokenSucceeds() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult));

    Task<String> task = firebaseAuthCredentialsProvider.getToken();

    assertThat(task.isComplete()).isTrue();
    assertThat(task.isSuccessful()).isTrue();
    assertThat(task.getResult()).isEqualTo("TestToken");
  }

  @Test
  public void getTokenShouldNotForceRefreshTheTokenIfInvalidateTokenIsNotCalled() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult));

    firebaseAuthCredentialsProvider.getToken();

    verify(mockInternalAuthProvider).getAccessToken(false);
  }

  @Test
  public void getTokenShouldRecursivelyCallItselfIfTheTokenCounterChanges() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    verify(mockInternalAuthProvider).addIdTokenListener(idTokenListenerCaptor.capture());
    when(mockGetTokenResult2.getToken()).thenReturn("TestToken2");
    TaskCompletionSource<GetTokenResult> getAccessTokenTaskCompletionSource =
        new TaskCompletionSource<>();

    // This delicate dance ensures the following order of operations:
    // 1. getToken() calls internalAuthProvider.getAccessToken(), which returns an uncompleted task.
    // 2. IdTokenListener.onIdTokenChanged() is invoked, which increments tokenCounter.
    // 3. The uncompleted task is completed.
    // 4. The continuation added to the task is invoked and detects that tokenCounter changed, and
    //       calls getToken() again.
    // 5. The recursively-called getToken() calls calls internalAuthProvider.getAccessToken(), which
    //       returns a completed task with the token "TestToken2".
    // 6. The task returned from the original getToken() invocation completes.
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(getAccessTokenTaskCompletionSource.getTask());
    Task<String> task = firebaseAuthCredentialsProvider.getToken();
    verify(mockInternalAuthProvider).getAccessToken(anyBoolean());
    idTokenListenerCaptor.getValue().onIdTokenChanged(mockInternalTokenResult);
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult2));
    getAccessTokenTaskCompletionSource.setResult(mockGetTokenResult);

    assertThat(task.isComplete()).isTrue();
    assertThat(task.isSuccessful()).isTrue();
    assertThat(task.getResult()).isEqualTo("TestToken2");
  }

  @Test
  public void invalidateTokenShouldCauseGetTokenToForceRefresh() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult));

    firebaseAuthCredentialsProvider.invalidateToken();
    firebaseAuthCredentialsProvider.getToken();

    verify(mockInternalAuthProvider).getAccessToken(true);
  }

  @Test
  public void invalidateTokenShouldOnlyForceRefreshOnTheImmediatelyFollowingGetTokenInvocation() {
    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
        new FirebaseAuthCredentialsProvider(new ImmediateDeferred<>(mockInternalAuthProvider));
    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult));

    firebaseAuthCredentialsProvider.invalidateToken();
    firebaseAuthCredentialsProvider.getToken();
    firebaseAuthCredentialsProvider.getToken();

    verify(mockInternalAuthProvider).getAccessToken(true);
    verify(mockInternalAuthProvider).getAccessToken(false);
  }
}

File: firebase-dataconnect/src/test/kotlin/com/google/firebase/dataconnect/auth/FirebaseAuthCredentialsProviderTest.kt
package com.google.firebase.dataconnect.auth

import com.google.android.gms.tasks.Tasks
import com.google.common.truth.Truth
import com.google.firebase.auth.GetTokenResult
import com.google.firebase.auth.internal.IdTokenListener
import com.google.firebase.auth.internal.InternalAuthProvider
import com.google.firebase.dataconnect.testutil.DelayedDeferred
import com.google.firebase.dataconnect.testutil.ImmediateDeferred
import com.google.firebase.dataconnect.testutil.UnavailableDeferred
import com.google.firebase.dataconnect.util.Listener
import com.google.firebase.internal.InternalTokenResult
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.runTest
import org.junit.Before
import org.junit.Test
import org.mockito.*
import org.mockito.Mockito.*

@Suppress("ReplaceCallWithBinaryOperator")
class FirebaseAuthCredentialsProviderTest {

  @Mock(name = "mockInternalAuthProvider")
  private lateinit var mockInternalAuthProvider: InternalAuthProvider

  @Mock(name = "mockUserListener") private lateinit var mockUserListener: Listener<User>

  @Captor private lateinit var idTokenListenerCaptor: ArgumentCaptor<IdTokenListener>

  @Mock(name = "mockInternalTokenResult")
  private lateinit var mockInternalTokenResult: InternalTokenResult

  @Mock private lateinit var mockGetTokenResult: GetTokenResult

  @Mock private lateinit var mockGetTokenResult2: GetTokenResult

  private var testScope = TestScope()

  @Before
  fun prepare() {
    MockitoAnnotations.initMocks(this)
  }

  @Test
  fun `setChangeListener should be called with unauthenticated if provider is not available`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider = FirebaseAuthCredentialsProvider(UnavailableDeferred())
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User.UNAUTHENTICATED)
    }
  }

  @Test
  fun `setChangeListener should be called with unauthenticated user if uid is null`() {
    testScope.runTest {
      `when`(mockInternalAuthProvider.uid).thenReturn(null)
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User.UNAUTHENTICATED)
    }
  }

  @Test
  fun `setChangeListener should be called with authenticated user if uid is not null`() {
    testScope.runTest {
      `when`(mockInternalAuthProvider.uid).thenReturn("TestUID")
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User("TestUID"))
    }
  }

  @Test
  fun `setChangeListener should be called with unauthenticated user when provider with null uid becomes available`() {
    testScope.runTest {
      val delayedDeferredInternalAuthProvider = DelayedDeferred<InternalAuthProvider>()
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider)
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User.UNAUTHENTICATED)
      `when`(mockInternalAuthProvider.uid).thenReturn(null)
      delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider)
      verify(mockUserListener, times(2)).onValue(User.UNAUTHENTICATED)
    }
  }

  @Test
  fun `setChangeListener should be called with authenticated user when provider with authenticated user becomes available`() {
    testScope.runTest {
      val delayedDeferredInternalAuthProvider = DelayedDeferred<InternalAuthProvider>()
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider)
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User.UNAUTHENTICATED)
      `when`(mockInternalAuthProvider.uid).thenReturn("TestUID")
      delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider)
      verify(mockUserListener).onValue(User("TestUID"))
    }
  }

  @Test
  fun `setChangeListener should be called when IdTokenChanges`() {
    testScope.runTest {
      `when`(mockInternalAuthProvider.uid).thenReturn("TestUID1")
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      verify(mockInternalAuthProvider).addIdTokenListener(idTokenListenerCaptor.capture())
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User("TestUID1"))
      `when`(mockInternalAuthProvider.uid).thenReturn("TestUID2")
      idTokenListenerCaptor.value.onIdTokenChanged(mockInternalTokenResult)
      verify(mockUserListener).onValue(User("TestUID2"))
    }
  }

  @Test
  fun `removeChangeListener should stop notifying the listener`() {
    testScope.runTest {
      val delayedDeferredInternalAuthProvider = DelayedDeferred<InternalAuthProvider>()
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(delayedDeferredInternalAuthProvider)
      firebaseAuthCredentialsProvider.setChangeListener(mockUserListener)
      verify(mockUserListener).onValue(User.UNAUTHENTICATED)
      firebaseAuthCredentialsProvider.removeChangeListener()
      delayedDeferredInternalAuthProvider.setInstance(mockInternalAuthProvider)
      verifyNoMoreInteractions(mockUserListener)
    }
  }

  @Test
  fun `removeChangeListener should not throw if provider is not available`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider = FirebaseAuthCredentialsProvider(UnavailableDeferred())
      firebaseAuthCredentialsProvider.removeChangeListener()
    }
  }

  @Test
  fun `removeChangeListener should unregister the idTokenListener`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      verify(mockInternalAuthProvider).addIdTokenListener(idTokenListenerCaptor.capture())
      firebaseAuthCredentialsProvider.removeChangeListener()
      verify(mockInternalAuthProvider).removeIdTokenListener(idTokenListenerCaptor.value)
    }
  }

  @Test
  fun `token should throw if provider is not available`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider = FirebaseAuthCredentialsProvider(UnavailableDeferred())
      val token = firebaseAuthCredentialsProvider.getToken()
      Truth.assertThat(token).isEqualTo(null)
    }
  }

  @Test
  fun `token should throw if access token fails`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      val getAccessTokenException = Exception()
      `when`(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forException(getAccessTokenException))
      val token = firebaseAuthCredentialsProvider.getToken()
      Truth.assertThat(token).isEqualTo(null)
    }
  }

  @Test
  fun `token should return a token string if getAccessToken succeeds`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      `when`(mockGetTokenResult.token).thenReturn("TestToken")
      `when`(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult))
      val token = firebaseAuthCredentialsProvider.getToken()
      Truth.assertThat(token).isEqualTo("TestToken")
    }
  }

  @Test
  fun `token should not force refresh if invalidateToken is not called`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      `when`(mockGetTokenResult.token).thenReturn("TestToken")
      `when`(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult))
      firebaseAuthCredentialsProvider.getToken()
      verify(mockInternalAuthProvider).getAccessToken(false)
    }
  }

  @Test
  fun `invalidate token should cause token force refresh`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      `when`(mockGetTokenResult.token).thenReturn("TestToken")
      `when`(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult))
      firebaseAuthCredentialsProvider.invalidateToken()
      firebaseAuthCredentialsProvider.getToken()
      verify(mockInternalAuthProvider).getAccessToken(true)
    }
  }

  @Test
  fun `invalidateToken should only force refresh on the immediately following getToken invocation`() {
    testScope.runTest {
      val firebaseAuthCredentialsProvider =
        FirebaseAuthCredentialsProvider(ImmediateDeferred(mockInternalAuthProvider))
      `when`(mockGetTokenResult.token).thenReturn("TestToken")
      `when`(mockInternalAuthProvider.getAccessToken(anyBoolean()))
        .thenReturn(Tasks.forResult(mockGetTokenResult))
      firebaseAuthCredentialsProvider.invalidateToken()
      firebaseAuthCredentialsProvider.getToken()
      firebaseAuthCredentialsProvider.getToken()
      verify(mockInternalAuthProvider).getAccessToken(true)
      verify(mockInternalAuthProvider).getAccessToken(false)
    }
  }
}
